### PR TITLE
automatically unzip before upload, re-zip any zipped files after Sentry artifact upload

### LINKED ISSUE #26: If used with ember-cli-deploy-gzip the uploaded files won't work.
Hi, I'm using this addon alongside `ember-cli-deploy-gzip` and since gzip runs first, the files deploy-sentry uploads to sentry are gzipped and won't work properly.

I tried disabling gzip for map files but since the js files are still being gzipped (as I want them to be on my server) this is what I see on my sentry stacktrace:

screen shot 2016-11-03 at 21 58 24

Any idea on how to solve this? Is it possible to check if the file is gzipped before uploading and if so?

1. Add a header to the file upload post which will make sentry properly handle the gzipped file?
2. Or unzip the file before uploading?

Thank you very much in advance

### PR DESCRIPTION
**Pull request by jasonmit**
I am opening this for a discussion, not necessarily to merge (especially the README changes!).

For now I'll continue to maintain this fork.

Fixes #26

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,63 +1,102 @@
...
+        let index = fs.readFileSync(indexPath, 'utf8');
+        index = index.replace(
+          '<meta name="sentry:revision">',
+          '<meta name="sentry:revision" content="' + revisionKey + '">'
+        );
         fs.writeFileSync(indexPath, index);
       },
 
-      upload: function(/* context */) {
+      upload(/* context */) {
```
Comment (jasonmit) [inline index.js:76]: This method is all that really changed, focus the review here.  Everything else is formatting and will be discarded when I open up the "official" PR if there is interest.

**Comment by duizendnegen** [comment]
OK - I know I've maintained for a long time that I wouldn't be supportive of this change, but seems like Sentry isn't moving on this topic... so let's try and get this landed & released.
Can we hide this behind a configuration settings, e.g. `unzipBeforeUploading: true`?

**Comment by jasonmit** [comment]
@duizendnegen sounds good.  The current implementation is only re-zip files that have been zipped.  I'll adjust it though to also be behind a flag at some point this week.

**Comment by kozak** [comment]
@jasonmit Useful stuff! Would you mind sharing if your branch was working fine ? 
I am asking because this PR is closed, and this might indicate you have found some problems with the implementation ?
Cheers!

**Comment by jasonmit** [comment]
@kozak branch is working great, we use it at Netflix as part of one of our deployments without issue for a few months now.  Unfortunately, I can't recall why I closed it now...

**Comment by dschmidt** [comment]
I'm sorry about the situation - to be honest I haven't done any Ember development in quite some time, that makes it really hard to maintain this project. 

Just for my understanding: This plugin is currently broken if `e-cli-d-gzip` is enabled and `keep` is disabled?

I am worried about the re-zipping part, especially because it doesn't neccessarily use the same zipping algorithm, see e.g. `zopfli` option of `e-cli-d-gzip`. 
Can't we unzip to a temp dir and upload from there (or directly upload the buffer) to leave the original gzipped files untouched?

If we make it opt in, it shouldn't break anything for anyone - so I'm even okay with the proposed solution here, but please mention the rezipping in the README.
Is it possible to access `ember-cli-deploy-gzip`'s configuration?  We could check `ember-cli-deploy-gzip`'s `keep` and only rezip/unzip if it's disabled and otherwise filter .gz files. Dunno if it's a good idea trying to be too smart - but I assume that would give the best out of the box experience.

Again I'm very sorry for the inconvenience and hope we can solve this issue for good finally.

**Comment by jasonmit** [comment]
> Just for my understanding: This plugin is currently broken if e-cli-d-gzip is enabled and keep is disabled?

I believe it's broken if you're using e-cli-d-gzip.  You end up uploading gzipped assets to sentry which sentry does not yet support.  Accepts the files but the callstack in the evented messages are unreadable blobs of text.

> Can't we unzip to a temp dir and upload from there (or directly upload the buffer) to leave the original gzipped files untouched?

I'm willing to can tackle this.

> Again I'm very sorry for the inconvenience

Don't be and thank you for this great addon!

**Comment by dschmidt** [comment]
Awesome!

**Comment by kozak** [comment]
Thank you both. @dschmidt is correct in that enabling: `keep: true` for the gzip plugin helps alleviate the issue. We ended up removing the gzip plugin and enabling gzipping on cloudfront.

**Comment by dschmidt** [comment]
Glad you found a solution that works for you!

**Comment by simonihmig** [comment]
Cross posting #52 here as another attempt to fix the issue. Feedback appreciated, maybe I'm missing something obvious...

**Pull request closed without merging**